src/System.Management.Automation/DscSupport/JsonCimDSCParser.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.IO;
+using System.Management.Automation;
+using System.Security;
+
+namespace Microsoft.PowerShell.DesiredStateConfiguration.Internal.Json
+{
+    /// <summary>
+    /// Class that does high level Cim schema parsing.
+    /// </summary>
+    internal class CimDSCParser
+    {
+        private readonly JsonDeserializer _json_deserializer;
+        
+        internal CimDSCParser()
+        {
+            _json_deserializer = JsonDeserializer.Create();
+        }
+
+        internal IEnumerable<PSObject> ParseSchemaJson(string filePath, bool useNewRunspace = false)
+        {
+            try
+            {
+                string json = File.ReadAllText(filePath);
+                string fileNameDefiningClass = Path.GetFileNameWithoutExtension(filePath);
+                int dotIndex = fileNameDefiningClass.IndexOf(".schema", StringComparison.InvariantCultureIgnoreCase);
+                if (dotIndex != -1)
+                {
+                    fileNameDefiningClass = fileNameDefiningClass.Substring(0, dotIndex);
+                }
+
+                var result = _json_deserializer.DeserializeClasses(json, useNewRunspace);
+                foreach (dynamic classObject in result)
+                {
+                    string superClassName = classObject.SuperClassName;
+                    string className = classObject.ClassName;
+                    if (superClassName?.Equals("OMI_BaseResource", StringComparison.OrdinalIgnoreCase) ?? false)
+                    {
+                        // Get the name of the file without schema.mof/json extension
+                        if (!(className.Equals(fileNameDefiningClass, StringComparison.OrdinalIgnoreCase)))
```

Comment:
What if `className` is `null`?

Reply:
Then the universe, as we know it, will cease to exist. :)

The processing of the file will stop with DeserializationError and that is probably the right thing to do considering that no further dsc operations will be possible if ClassName is not specified. Though the chances of this are rather low and we are also expecting to have a schema validation cmdlet to check for these sort of things.